This notebook is related to a growth model, which is an attempt to identify drivers and levers of growth

In [1]:
import numpy as np
import altair as alt

import pydqt
import pandas as pd
import os
import itertools
from importlib import reload
import numpy as np

# os.sys.path.append('../timeseries') 
# from tsx import sum_of_nchoosek as snck
# import tsx as tsx

pydqt.set_workspace(root='/Users/mingham/research/workspaces/',name='growth_model')
pydqt.set_schema('core_wip')

In [10]:
q=pydqt.Query('test.sql')
q.run()
sum(q.df.duplicated())

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJfb9owFMW%2FSuQ9J3ZcBqtFQBkILRJlEYRK9M2LTWrVsVP%2FIfTbz4QidQ%2BttAdLln2O7%2B%2F63On83MroxI0VWmUgTRCIuKo1E6rJwL5axT9AZB1VjEqteAbeuAXz2dTSVnYk9%2B5Zbfmr59ZF4SFlyXCRAW8U0dQKSxRtuSWuJrv8YU1wgkhntNO1luCD5WsHtZYbFwhvFmZFwHt2riMQ9n2f9HeJNg3ECCGI7mFQXSTfbvpz6OkTfQrR6KIPiiAv39l%2BCnX9gq%2Bw%2FlxFlvyqqjIuf%2B8qEOU31IVW1rfc7Lg5iZrvt%2BsrgA0Eq914co9xwn3ch6%2BL08Qq3R8lfeG1bjvvwqtJ2MEjZ1DqRoTGi2UGuhfBjrTd68XCv1brbS5P28K15Y4e8vVkQ8%2FL9YluPROHw6YoHmoQPd6SxZdkC2s9L9QlTxeOEB7FCMf4rkoxSVPyfZTgMXoC0TJACUXd4LxB07rWXjmbNFo3kg98Gl7Cw1Cwbh6WYNkCpY1ik%2FEZXIeEDDXN7H9bn8KP7vd524QIimWppajfopU2LXWfJ5Qm6XAiWHwcpIS3VMicMcOtDUlJqfuF4dSFsXbGcwBn16r%2FDvbsLw%3D%3D&RelayState=58335 to authenticate...


/Users/mingham/research/.venv/lib/python3.11/site-packages/pydqt/pydqt.py:633: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(self.sql.text, conn)


1058

555 duplicates for 'WEEK'
1058 duplicates for 'MONTH'
1058 duplicates for 'QUARTER'



In [2]:
from itertools import combinations
x = combinations(['a','b','c'],2)

In [8]:
df=pd.read_csv('~/Downloads/dups.csv')

In [15]:
tf=df.duplicated('NC_COMPONENTS_DERIVED_FROM_BASE_ID',False)

In [17]:
df[tf==True].sort_values(['NC_COMPONENTS_DERIVED_FROM_BASE_ID'])

NC_COMPONENTS_DERIVED_FROM_BASE_ID PERIOD_DS_FREQ  \
13668                MONTH|2024-02-01|ALL|ALL|ALL|ALL          MONTH   
23601                MONTH|2024-02-01|ALL|ALL|ALL|ALL          MONTH   
73341                MONTH|2024-02-01|ALL|ALL|ALL|ALL          MONTH   
23602                MONTH|2024-02-01|ALL|ALL|ALL|ALL          MONTH   
58358                MONTH|2024-02-01|ALL|ALL|ALL|ALL          MONTH   
...                                               ...            ...   
45973  WEEK|2024-02-12|null|TRUE|web only signup|PAID           WEEK   
60961  WEEK|2024-02-12|null|TRUE|web only signup|PAID           WEEK   
60920  WEEK|2024-02-12|null|TRUE|web only signup|PAID           WEEK   
31076  WEEK|2024-02-12|null|TRUE|web only signup|PAID           WEEK   
65937  WEEK|2024-02-12|null|TRUE|web only signup|PAID           WEEK   

         EVENT_DS MAJOR_MARKET_PARTNER IS_MEMBER        USER_TYPE  \
13668  2024-02-01                  ALL       ALL              ALL   
23601  2024-02-01                  ALL       ALL              ALL   
73341  2024-02-01                  ALL       ALL              ALL   
23602  2024-02-01                  ALL       ALL              ALL   
58358  2024-02-01                  ALL       ALL              ALL   
...           ...                  ...       ...              ...   
45973  2024-02-12                  NaN      TRUE  web only signup   
60961  2024-02-12                  NaN      TRUE  web only signup   
60920  2024-02-12                  NaN      TRUE  web only signup   
31076  2024-02-12                  NaN      TRUE  web only signup   
65937  2024-02-12                  NaN      TRUE  web only signup   

      SESSION_TRAFFIC_SOURCE_GROUPING  USER_COUNT  SESSION_COUNT  \
13668                             ALL      210935         225764   
23601                             ALL      115697         124483   
73341                             ALL      215116         230979   
23602                             ALL      211589         226163   
58358                             ALL      533542         581169   
...                               ...         ...            ...   
45973                            PAID           4              4   
60961                            PAID           2              2   
60920                            PAID           2              2   
31076                            PAID           1              1   
65937                            PAID           2              2   

       SESSIONS_PER_USER  ...  CONVERSION_COUNT  TRACK_CONVERSION_RATE  \
13668           1.070301  ...            1139.0               0.102465   
23601           1.075940  ...             605.0               0.070835   
73341           1.073742  ...             839.0               0.075928   
23602           1.068879  ...            1322.0               0.115691   
58358           1.089266  ...            4122.0               0.009554   
...                  ...  ...               ...                    ...   
45973           1.000000  ...               0.0               0.000000   
60961           1.000000  ...               NaN                    NaN   
60920           1.000000  ...               NaN                    NaN   
31076           1.000000  ...               NaN                    NaN   
65937           1.000000  ...               NaN                    NaN   

       ORDER_COUNT  ORDER_COUNT_PER_CONVERSION            GMV         AOV  \
13668       1366.0                    1.199298  325084.577804  237.982853   
23601        823.0                    1.360331  205783.629618  250.040862   
73341        999.0                    1.190703  221732.470634  221.954425   
23602       1540.0                    1.164902  374864.230629  243.418332   
58358       4221.0                    1.024017  855621.211718  202.705807   
...            ...                         ...            ...         ...   
45973          0.0                         NaN            NaN         NaN   
609

This notebook is investigating why AOV and Customer Purchase Rate is down in FY24.  I shall be using data from CORE_WIP.nc_components_derived_aggregates_v2 (see template for details)

In [2]:
# q=pydqt.Query('dev.sql')
q=pydqt.Query('nc_components_derived_aggregates_from_base.sql',min_query_date='2023-11-01')
q.run(database='lyst_dev',schema='lyst_analytics')

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJRT9swFIX%2FSuQ9J3ZMacFqWqWt2lWCraOBaXsziRMsHDvzdUj593OaVoIHkHiIFNn3%2BDv3njudH2oVvAgL0ugExRFBgdC5KaSuEnSfrcMrFIDjuuDKaJGgVwFoPpsCr1XD0tY96TvxrxXgAv%2BQBtZfJKi1mhkOEpjmtQDmcrZPb28YjQjjAMI6j0MnSQHSs56caxjGXddF3UVkbIUpIQSTa%2Byr%2BpJv6A2i%2BZzRWONMbtRZcvA9fYCIMRn1CF%2FhCbuTcCH1MILPKI9DEbDvWbYLdz%2F3GQrSc3dLo6Gthd0L%2ByJzcX93MxgA72C9H0%2BuKY1EG3Z%2BdGEcgTZdqfizyE3dtM6%2FGvk%2FXIoCK1NJP6vtKkHNsyyax80fvQElb9PLbfa77lrYHKReXHHaHB4WZdcsy%2FGmTn%2FxUY6Ch3OytE92C9CKre7zdP6I0FFIaEhJFo%2FZ6JLRcTSZXPxFwcqbkpq7o%2FJsmue5abWDqDKmUuLoz%2BA%2BDIq9rbn%2FZJEsSVzpYjI%2BoGFJ2JFpZ19tfYrfqk%2F79sNHsF3tjJL5a7A2tubu44TiKD6eyCIsj6VM1FyqtCisAPBJKWW6pRXc%2BbV2thUIzwbq%2B8We%2FQc%3D&RelayState=55467 t

/Users/mingham/research/.venv/lib/python3.11/site-packages/pydqt/pydqt.py:633: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(self.sql.text, conn)


PERIOD_DS_FREQ   EVENT_DS     MAJOR_MARKET_PARTNER IS_MEMBER  \
0                 DAY 2023-12-30                      ALL      TRUE   
1                 DAY 2024-01-11                      ALL      TRUE   
2                 DAY 2023-11-03                      ALL      TRUE   
3                 DAY 2023-12-06                      ALL      TRUE   
4                 DAY 2023-11-01                      ALL      TRUE   
...               ...        ...                      ...       ...   
377515        QUARTER 2024-01-01            Latin America       ALL   
377516        QUARTER 2024-01-01                      ROW       ALL   
377517        QUARTER 2024-01-01  Europe (rest of Europe)       ALL   
377518        QUARTER 2023-10-01                       IT       ALL   
377519        QUARTER 2023-10-01                       DE       ALL   

        USER_TYPE SESSION_TRAFFIC_SOURCE_GROUPING  USER_COUNT  SESSION_COUNT  \
0       app first                          PLA-EC         630            714   
1       app first                          DIRECT       55750          74069   
2       app first                            PAID        2753           2949   
3             w2a                            PAID        1061           1165   
4             w2a                         ORGANIC        8309          10326   
...           ...                             ...         ...            ...   
377515        ALL                             ALL       18095          20668   
377516        ALL                             ALL        3329           3800   
377517        ALL                             ALL       62758          70894   
377518        ALL                             ALL       32764          34727   
377519        ALL                             ALL       65761          69215   

        SESSIONS_PER_USER  LEAD_GENERATION_COUNT  ...  CONVERSION_COUNT  \
0                1.133333                    162  ...               9.0   
1                1.328592                  12684  ...             212.0   
2                1.071195                    715  ...               6.0   
3                1.098021                    727  ...              12.0   
4                1.242749                   4252  ...              64.0   
...                   ...                    ...  ...               ...   
377515           1.142194                   7839  ...              53.0   
377516           1.141484                   1773  ...              10.0   
377517           1.129641                  48470  ...             276.0   
377518           1.059913                  35750  ...             298.0   
377519           1.052524                  70044  ...            1464.0   

        TRACK_CONVERSION_RATE  ORDER_COUNT  ORDER_COUNT_PER_CONVERSION  \
0                    0.055556         10.0                    1.111111   
1                    0.016714        218.0                    1.028302   
2                    0.008392          6.0                    1.000000   
3                    0.016506         12.0                    1.000000   
4                    0.015052         78.0                    1.218750   
...                       ...          ...                         ...   
377515               0.006761         59.0                    1.113208   
377516               0.005640         10.0                    1.000000   
377517               0.005694        277.0                    1.003623   
377518               0.008336        321.0                    1.077181   
377519               0.020901       1571.0                    1.073087   

                  GMV         AOV  GROSS_COMMISSION  NET_COMMISSION  CPA_RATE  \
0         5650.447280  565.044728        666.597200      666.677200  0.117972   
1        70564.934878  323.692362      10291.141754    10203.807638  0.145839   
2          969.169597  161.528266        152.231023      138.431023  0.157074   
3         5834.368131  486.197344        757.832127      748.642127  0.129891  

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJRb9sgFIX%2FisWebTBLkwbFidxkUSNla9a4U9c3aoiLisHj4jjtrx9xGql7aKU9ICE4h%2Ftdzp3MDrWO9tKBsiZDaUJQJE1phTJVhu6KZXyJIvDcCK6tkRl6kYBm0wnwWjcsb%2F2TuZV%2FWgk%2BCg8ZYP1FhlpnmOWggBleS2C%2BZNv8%2B5rRhLDGWW9Lq9E7y%2BcODiCdD4RniwAV8J68bxjGXdcl3dfEugpTQggmYxxUR8mXs%2F4QevpAn2IyOOqDIsg3b2xXypy%2B4DOsx5MI2HVRbOLNzbZAUX5GnVsDbS3dVrq9KuXd7foEAIFguR2OxpQmso278HVxmoCx3U7zZ1nauml9eDUJO7yTAmtbqdD4apGh5lmJkdvsyqvfP1NaD8V6DXx%2F8%2B3%2B9VU%2BpGpIXS5WzeNwfn99MTZ5iaJf52TpMdkVQCtX5pinD0eEDmJCY5oW6YCRS0YuklFKH1C0CFDKcN87z9C8LG1rPCSVtZWWPZ%2FFx%2FAoVqKZhaVENidpZcRoeECnIWF9TTf939Yn%2BL37bd5%2BhAhWi43VqnyJltbV3H%2BcUJqk%2FYkS8a6XMllzpXMhnAQISWltu7mT3Iex9q6VCE9PVf8d7Olf&RelayState=58463 to authenticate..

In [23]:
tf=q.df.duplicated(['PERIOD_DS_FREQ','EVENT_DS','MAJOR_MARKET_PARTNER','IS_MEMBER','USER_TYPE','SESSION_TRAFFIC_SOURCE_GROUPING'],False)

In [24]:
sum(tf)

281298

In [28]:
dups_df=q.df[tf].sort_values(['PERIOD_DS_FREQ','EVENT_DS','MAJOR_MARKET_PARTNER','IS_MEMBER','USER_TYPE','SESSION_TRAFFIC_SOURCE_GROUPING'])

In [30]:
dups_df[dups_df['PERIOD_DS_FREQ']=='QUARTER']

PERIOD_DS_FREQ   EVENT_DS MAJOR_MARKET_PARTNER IS_MEMBER  \
6235          QUARTER 2023-10-01                  ALL       ALL   
17707         QUARTER 2023-10-01                  ALL       ALL   
23359         QUARTER 2023-10-01                  ALL       ALL   
29374         QUARTER 2023-10-01                  ALL       ALL   
41185         QUARTER 2023-10-01                  ALL       ALL   
...               ...        ...                  ...       ...   
335189        QUARTER 2024-01-01                 None      TRUE   
340909        QUARTER 2024-01-01                 None      TRUE   
352940        QUARTER 2024-01-01                 None      TRUE   
371030        QUARTER 2024-01-01                 None      TRUE   
371108        QUARTER 2024-01-01                 None      TRUE   

              USER_TYPE SESSION_TRAFFIC_SOURCE_GROUPING  USER_COUNT  \
6235                ALL                             ALL      711064   
17707               ALL                             ALL      612851   
23359               ALL                             ALL      657206   
29374               ALL                             ALL      935651   
41185               ALL                             ALL      691237   
...                 ...                             ...         ...   
335189  web only signup                             ALL          16   
340909  web only signup                             ALL          17   
352940  web only signup                             ALL          18   
371030  web only signup                             ALL          30   
371108  web only signup                             ALL          18   

        SESSION_COUNT  SESSIONS_PER_USER  LEAD_GENERATION_COUNT  ...  \
6235           775732           1.090945                 571145  ...   
17707          668316           1.090503                 475725  ...   
23359          716419           1.090098                 560588  ...   
29374         1025007           1.095501                 810472  ...   
41185          756344           1.094189                 589829  ...   
...               ...                ...                    ...  ...   
335189             21           1.312500                     43  ...   
340909             21           1.235294                     42  ...   
352940             23           1.277778                     33  ...   
371030             36           1.200000                     56  ...   
371108             19           1.055556                     34  ...   

        CONVERSION_COUNT  TRACK_CONVERSION_RATE  ORDER_COUNT  \
6235              9103.0               0.015938       9775.0   
17707             7372.0               0.015496       7920.0   
23359            10321.0               0.018411      11062.0   
29374            15691.0               0.019360      16792.0   
41185             9860.0               0.016717      10532.0   
...                  ...                    ...          ...   
335189               1.0               0.023256          1.0   
340909               0.0               0.000000          0.0   
352940               0.0               0.000000          0.0   
371030               0.0               0.000000          0.0   
371108               0.0               0.000000          0.0   

        ORDER_COUNT_PER_CONVERSION           GMV         AOV  \
6235                      1.073822  2.165478e+06  221.532322   
17707                     1.074335  1.814964e+06  229.162123   
23359                     1.071795  2.126388e+06  192.224519   
29374                     1.070168  3.316495e+06  197.504451   
41185                     1.068154  2.189463e+06  207.886717   
...                            ...           ...         ...   
335189                    1.000000  2.376700e+02  237.670000   
340909                         NaN           NaN         NaN   
352940                         NaN           NaN         NaN   
371030                         NaN           NaN         NaN   
371108     

In [27]:
q.sql.open()

In [9]:
q.df=df

In [12]:
q.df['MAJOR_MARKET_PARTNER']='BOO'

In [31]:
q=pydqt.Query('select * from nc_components_base')
q.run?

Signature: q.run(schema='')
Docstring: <no docstring>
File:      ~/research/.venv/lib/python3.11/site-packages/pydqt/pydqt.py
Type:      method

In [8]:
df['MAJOR_MARKET_PARTNER']='YO'

In [41]:
'a_b_c_s|w_x_y'.replace('_',' ').split('|')

['a b c s', 'w x y']

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJfb9owFMW%2FSuQ9J3ZSKK1FQBRUitYWVCja%2BubGJnh17MzXJnSffg5%2FpO6hlfYQKbLP8e%2Fee25%2FuK9UtBMWpNE5ShOCIqELw6Uuc%2FS8uo2vUASOac6U0SJH7wLQcNAHVqmajrzb6ifx2wtwUXhIA20vcuStpoaBBKpZJYC6gi5HD%2Fc0SwhlAMK6gEMnCwcZWFvnaopx0zRJc5EYW%2BKMEILJNQ6qVvINfUDUXzNqa5wpjDpb9qGnTxApJp0WERSBsDgZb6Q%2BjuAryutRBPRutVrEi%2FlyhaLRubux0eArYZfC7mQhnp%2FujwVAqOB2edm7zrJE%2BLgJo4vTBLRpNoq9icJUtXfh1ST84Y3gWJlShlnNJjmq3yS3bL1%2BEYubXw9Zdw6%2BN93%2FnG2nu%2B9342o6F12W%2Bj%2BFfp3%2BeEyhQNH6nGzWJjsD8GKm2zxdOCJZJyZZTLor0qOdHs0ukqtu%2BoKiSShKauYOznPRrCiM1w6S0phSiUN9BrdhZFjyehg%2ByfMxSUvNe5d7dFwSemDawf%2B23scf3ad9ewwRzCYLo2TxHt0aWzH3eUJpkh5OJI83BykVFZNqxLkVACEppUwztoK5sNbOeoHw4Ej9d7EHfwE%3D&RelayState=51927 to 

In [3]:
q.run(schema='lyst_analytics')

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJfb9owFMW%2FSuQ9J04CK8UiIApDZaUdK3TS%2BubFl9TCsTNfh0A%2FfR3%2BSN1DK01KpMg5179z77mD0b5UwQ4sSqMzkkQxCUDnRkhdZORpPQuvSYCOa8GV0ZCRAyAZDQfIS1Wxce1e9CP8rQFd4C%2FSyNofGamtZoajRKZ5Cchczlbj%2BwVLo5hxRLDO48i5RKD0rBfnKkZp0zRR04mMLWgaxzGN%2B9SrWskX8g5Rfc6orHEmN%2BpSsvc9fYBIaNxtEV7hCctz4Y3UpxF8RvlzEiG7Xa%2BX4fLHak2C8aW7idFYl2BXYHcyh6fHxckAegez1VWvn6YR1GHjRxcmEWrTbBTfQm7Kqnb%2B1sh%2F0Q0Iqkwh%2Fazm04xUWymKii%2Bev7%2Fe3N%2B5%2Bev256GHW%2F7t60Onqm53sJMN3u36arbQ3envnAS%2FLsmmbbJzxBrmus3T%2BaM47YZx6p91krJOn3Wuo15y9UyCqTclNXfHyotpnuem1g6jwphCwdGfoW0YKZWiGvlXimwSJ4UWvas9OS0JOzLt8H9bH9D31ed9e%2FARzKdLo2R%2BCGbGltx9nFASJccTKcLNUcqg5FKNhbCA6JNSyjQTC9z5tXa2BkKHJ%2Bq%2Fiz18Aw%3D%3D&RelayState=62357

/Users/mingham/research/.venv/lib/python3.11/site-packages/pydqt/pydqt.py:629: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(self.sql.text, conn)


PERIOD_DS_FREQ   EVENT_DS     MAJOR_MARKET_PARTNER  IS_MEMBER  USER_TYPE  \
0              WEEK 2023-11-06                       IT      False   anon web   
1              WEEK 2023-11-06  Europe (rest of Europe)      False   anon web   
2              WEEK 2023-11-06                       IT       True  app first   
3              WEEK 2023-11-13                       ES       True  app first   
4              WEEK 2023-12-18  Europe (rest of Europe)       True  app first   
...             ...        ...                      ...        ...        ...   
4189          MONTH 2023-12-01              Middle East       True  app first   
4190          MONTH 2023-12-01                       NL       True  app first   
4191          MONTH 2023-12-01                     None       True        w2a   
4192          MONTH 2023-12-01                     Asia      False   anon web   
4193          MONTH 2023-12-01                       AU      False   anon web   

     SESSION_TRAFFIC_SOURCE_GROUPING  USER_COUNT  SESSION_COUNT  \
0                            ORGANIC       66271          74277   
1                             DIRECT       24805          28138   
2                            ORGANIC        1680           2218   
3                             DIRECT        4572           8724   
4                            ORGANIC       16294          24758   
...                              ...         ...            ...   
4189                            PAID        9914          12832   
4190                            PAID         152            202   
4191                          DIRECT         717           2305   
4192                             CSS        3089           3977   
4193                          PLA-CO          45             51   

      SESSIONS_PER_USER  LEAD_GENERATION_COUNT  ...  CONVERSION_COUNT  \
0              1.120807                  44548  ...             304.0   
1              1.134368                  17564  ...              85.0   
2              1.320238                    803  ...               5.0   
3              1.908136                   1697  ...              19.0   
4              1.519455                   6176  ...              87.0   
...                 ...                    ...  ...               ...   
4189           1.294331                   3354  ...              35.0   
4190           1.328947                    109  ...               0.0   
4191           3.214784                    709  ...              11.0   
4192           1.287472                   4094  ...              16.0   
4193           1.133333                      4  ...               0.0   

      TRACK_CONVERSION_RATE  ORDER_COUNT  ORDER_COUNT_PER_CONVERSION  \
0                  0.006824        337.0                    1.108553   
1                  0.004839         96.0                    1.129412   
2                  0.006227          5.0                    1.000000   
3                  0.011196         20.0                    1.052632   
4                  0.014087         89.0                    1.022989   
...                     ...          ...                         ...   
4189               0.010435         37.0                    1.057143   
4190               0.000000          0.0                         NaN   
4191               0.015515         12.0                    1.090909   
4192               0.003908         19.0                    1.187500   
4193               0.000000          0.0                         NaN   

             AOV  CPA_RATE  RETAIN_RATE           GMV  GROSS_COMMISSION  \
0     222.843297  0.139643     0.694203  75098.191061      10486.905925   
1     287.218986  0.135185     0.652135  27573.022617       3727.450898   
2     129.126446  0.144619     0.834048    645.632231         93.370742   
3     336.437141  0.133616     0.899361   6728.742810        899.066752   
4     319.559272  0.135480     0.954067  28440.775250       3853.154202   
...          ...       ...         

In [4]:
q.write_sql('nc_components_v2',schema='CORE_WIP')

Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...
Going to open: https://accounts.google.com/o/saml2/idp?idpid=C01gnd76x&SAMLRequest=lZJRb9owFIX%2FSuQ9J3YCK6tFQAyUwtRuDCiCvnmxSb06durrNHS%2Ffk4AqXtopUmJFDnn%2Bjv3njscH0sVvAgL0ugUxRFBgdC54VIXKbrfZOEXFIBjmjNltEjRqwA0Hg2Blaqik9o96pV4rgW4wF%2BkgbY%2FUlRbTQ0DCVSzUgB1OV1P7m5pEhHKAIR1HofOJRykZz06V1GMm6aJml5kbIETQggm19irWskn9AZRfcyorHEmN%2BpScvQ9vYOIMem3CK%2FwhOW58KvUpxF8RPl1EgGdbzbLcPljvUHB5NLd1GioS2HXwr7IXNyvbk8GwDvI1leD6ySJRB02fnRhHIE2zUGxJ5GbsqqdvzXyX%2FggOFamkH5Wi1mKqifJ93tI2Lb%2Fx9w8Z9tyvvt22LEH3Sun1V7I30VW3TC2W%2F3UxdzkKNhekk3aZBcAtVjoNk%2Fnj0jSD0nin02c0D6hcS%2BKPw8eUDDzpqRmrqu8mGZ5bmrtICqMKZTo%2FBnchpFgyauxfyVPpyQuNB9cHdFpSWjHtKP%2FbX2I31af9%2B27j2AxWxol89cgM7Zk7v2E4ijuTiQPD52UipJJNeHcCgCflFKmmVrBnF9rZ2uB8OhE%2FXexR38B&RelayState=62374 to authenticate...


In [28]:
def get_table_metadata(df):
    def map_dtypes(x):
        print(x)
        if (x == 'object') or (x=='category'):
            return 'VARCHAR'
        elif 'date' in x:
            return 'DATE'
        elif 'int' in x:
            return 'NUMERIC'  
        elif 'float' in x: return 'FLOAT' 
        else:
            print("cannot parse pandas dtype")
    sf_dtypes = [map_dtypes(str(s)) for s in df.dtypes]
    table_metadata = ", ". join([" ".join([y.upper(), x]) for x, y in zip(sf_dtypes, list(df.columns))])
    return table_metadata


In [30]:
q.df.dtypes

PERIOD_DS_FREQ                      object
EVENT_DS                            object
MAJOR_MARKET_PARTNER                object
IS_MEMBER                             bool
USER_TYPE                           object
SESSION_TRAFFIC_SOURCE_GROUPING     object
USER_COUNT                           int64
SESSION_COUNT                        int64
SESSIONS_PER_USER                  float64
LEAD_GENERATION_COUNT                int64
LEAD_GENERATION_RATE               float64
CONVERSION_COUNT                   float64
TRACK_CONVERSION_RATE              float64
ORDER_COUNT                        float64
ORDER_COUNT_PER_CONVERSION         float64
AOV                                float64
CPA_RATE                           float64
RETAIN_RATE                        float64
GMV                                float64
GROSS_COMMISSION                   float64
NET_COMMISSION                     float64
dtype: object

In [29]:
get_table_metadata(q.df)

object
object
object
bool
cannot parse pandas dtype
object
object
int64
int64
float64
int64
float64
float64
float64
float64
float64
float64
float64
float64
float64
float64
float64


TypeError: sequence item 1: expected str instance, NoneType found

In [25]:
# q.df.dtypes
for i in q.df['PERIOD_DS_FREQ']:
    if type(i)!=str:
        print(i,type(i))

# type(q.df['PERIOD_DS_FREQ'].iloc[0])

In [16]:
get_table_metadata(q.df)

cannot parse pandas dtype


TypeError: sequence item 1: expected str instance, NoneType found

In [11]:
q.df

PERIOD_DS_FREQ             EVENT_DS     MAJOR_MARKET_PARTNER  IS_MEMBER  \
0                DAY  2023-12-12:00-00-00                       DE      False   
1                DAY  2023-11-12:00-00-00                       GB      False   
2                DAY  2023-11-12:00-00-00                       US       True   
3                DAY  2023-11-12:00-00-00                       NL       True   
4                DAY  2023-11-28:00-00-00                       US       True   
...              ...                  ...                      ...        ...   
23645            DAY  2023-11-30:00-00-00                     None      False   
23646            DAY  2023-11-07:00-00-00                     None       True   
23647            DAY  2023-12-13:00-00-00              Middle East       True   
23648            DAY  2023-12-25:00-00-00  Europe (rest of Europe)       True   
23649            DAY  2023-11-19:00-00-00                       AU       True   

             USER_TYPE SESSION_TRAFFIC_SOURCE_GROUPING  USER_COUNT  \
0             anon web                             CSS        6597   
1             anon web                         ORGANIC       45173   
2      web only signup                         ORGANIC        2413   
3            app first                          DIRECT         387   
4            app first                            PAID         307   
...                ...                             ...         ...   
23645  no on lyst view                            PAID           2   
23646              w2a                          PLA-EC           1   
23647              w2a                        APP PAID           1   
23648  web only signup                          PLA-EC           1   
23649  web only signup                             CSS           1   

       SESSION_COUNT  SESSIONS_PER_USER  LEAD_GENERATION_COUNT  ...  \
0               7266           1.101410                  10220  ...   
1              47876           1.059837                  26122  ...   
2               2872           1.190220                   3535  ...   
3                561           1.449612                    168  ...   
4                318           1.035831                     59  ...   
...              ...                ...                    ...  ...   
23645              2           1.000000                      1  ...   
23646              1           1.000000                      0  ...   
23647              1           1.000000                      0  ...   
23648              1           1.000000                      5  ...   
23649              1           1.000000                      1  ...   

       CONVERSION_COUNT  TRACK_CONVERSION_RATE  ORDER_COUNT  \
0                 216.0               0.021135        231.0   
1                 406.0               0.015542        460.0   
2                  57.0               0.016124         62.0   
3                   3.0               0.017857          3.0   
4                   4.0               0.067797          4.0   
...                 ...                    ...          ...   
23645               0.0               0.000000          0.0   
23646               NaN                    NaN          NaN   
23647               NaN                    NaN          NaN   
23648               0.0               0.000000          0.0   
23649               0.0               0.000000          0.0   

       ORDER_COUNT_PER_CONVERSION         AOV  CPA_RATE  RETAIN_RATE  \
0                        1.069444  258.639873  0.150538     0.634315   
1                        1.133005  138.614336  0.137929     0.719278   
2                        1.087719  365.714918  0.143284     0.875257   
3                        1.000000  139.355467  0.146841     1.000000   
4                        1.000000  233.965000  0.115787     1.000000   
...                           ...         ...       ...          ...   
23645                         NaN         NaN       NaN          NaN   
23646 